src/sumo_store_mnesia.erl

```diff
@@ -0,0 +1,308 @@
+  parse(Options, [{ram_copies, [node()]} | Acc]);
+parse([{ram_copies, Nodes} | Options], Acc) ->
+  parse(Options, [{ram_copies, Nodes} | Acc]);
+parse([{majority, Flag} | Options], Acc) ->
+  parse(Options, [{majority, Flag} | Acc]);
+parse([{snmp, SnmpStruct} | Options], Acc) ->
+  parse(Options, [{snmp, SnmpStruct} | Acc]);
+parse([{storage_properties, Props} | Options], Acc) ->
+  parse(Options, [{storage_properties, Props} | Acc]);
+parse([_IgnoredOption | Options], Acc) ->
+  parse(Options, Acc).
+
+new_id(DocName, FieldType) ->
+  NewId = new_id(FieldType),
+  case mnesia:dirty_read(DocName, NewId) of
+    [] -> NewId;
+    _ -> new_id(DocName, FieldType)
+  end.
+
+new_id(string) -> uuid:uuid_to_string(uuid:get_v4(), standard);
+new_id(binary) -> uuid:uuid_to_string(uuid:get_v4(), binary_standard);
+new_id(text) -> uuid:uuid_to_string(uuid:get_v4(), binary_nodash);
+new_id(integer) -> <<Id:128>> = uuid:get_v4(), Id;
+new_id(float) -> <<Id:128>> = uuid:get_v4(), Id * 1.0;
+new_id(FieldType) -> throw({unimplemented, FieldType}).
+
+%% @ref http://www.erlang.org/doc/apps/erts/match_spec.html
+build_match_spec(DocName, Condition) when not is_list(Condition) ->
+  build_match_spec(DocName, [Condition]);
+build_match_spec(DocName, Conditions) ->
+  Schema = sumo_internal:get_schema(DocName),
+  Fields = schema_field_names(Schema),
+  FieldsMap =
+    maps:from_list(
+      [field_tuple(I, Fields) || I <- lists:seq(1, length(Fields))]),
+  MatchHead = list_to_tuple([DocName | lists:sort(maps:values(FieldsMap))]),
+  Guard =
+    [condition_to_guard(Condition, FieldsMap) || Condition <- Conditions] ,
+  Result = '$_',
+  [{MatchHead, Guard, [Result]}].
+
+field_tuple(I, Fields) ->
+  FieldName = lists:nth(I, Fields),
+  FieldWildcard =
+    list_to_atom([$$ | integer_to_list(I)]),
+  {FieldName, FieldWildcard}.
+
+condition_to_guard({'and', [Expr1]}, FieldsMap) ->
+  condition_to_guard(Expr1, FieldsMap);
+condition_to_guard({'and', [Expr1 | Exprs]}, FieldsMap) ->
+  { 'andalso'
+  , condition_to_guard(Expr1, FieldsMap)
+  , condition_to_guard({'and', Exprs}, FieldsMap)
+  };
+condition_to_guard({'or', [Expr1]}, FieldsMap) ->
+  condition_to_guard(Expr1, FieldsMap);
+condition_to_guard({'or', [Expr1 | Exprs]}, FieldsMap) ->
+  { 'orelse'
+  , condition_to_guard(Expr1, FieldsMap)
+  , condition_to_guard({'or', Exprs}, FieldsMap)
+  };
+condition_to_guard({'not', Expr}, FieldsMap) ->
+  {'not', condition_to_guard(Expr, FieldsMap)};
+condition_to_guard({Name1, Op, Name2}, FieldsMap) when is_atom(Name2) ->
+  check_operator(Op),
+  %NOTE: Name2 can be a field name or a value, that's why the following happens
+  {Op, maps:get(Name1, FieldsMap), maps:get(Name2, FieldsMap, Name2)};
+condition_to_guard({Name1, Op, Value}, FieldsMap) ->
+  check_operator(Op),
+  {Op, maps:get(Name1, FieldsMap), Value};
+condition_to_guard({Name, 'null'}, FieldsMap) ->
+  condition_to_guard({Name, '==', undefined}, FieldsMap);
```

Comment:
According to **Dialyzer**:
>  The call sumo_store_mnesia:condition_to_guard({_,'==','undefined'},FieldsMap:#{}) will never return since it differs in the 1st argument from the success typing arguments: ({'and',[{_,_} | {_,_,_},...]} | {'not',{'and',[{_,_} | {_,_,_},...]} | {'not',{'and',[any(),...]} | {'not',{_,_} | {_,_,_}} | {'or',[any(),...]} | {_,'<' | '>' | '>=',_}} | {'or',[{_,_} | {_,_,_},...]} | {_,'<' | '>' | '>=',_}} | {'or',[{_,_} | {_,_,_},...]} | {_,'<' | '>' | '>=',_},#{})